packages/web3-providers-ws/src/index.ts

```diff
@@ -0,0 +1,387 @@
+import { EventEmitter } from 'events';
+import {
+	Web3BaseProvider,
+	Web3BaseProviderStatus,
+	JsonRpcId,
+	JsonRpcPayload,
+	JsonRpcResponse,
+	JsonRpcResult,
+	Web3BaseProviderCallback,
+	ConnectionEvent,
+	SubscriptionResultNotification,
+} from 'web3-common';
+import { IMessageEvent, w3cwebsocket as W3WS } from 'websocket';
+import {
+	InvalidConnectionError,
+	ConnectionTimeoutError,
+	PendingRequestsOnReconnectingError,
+	ConnectionNotOpenError,
+	InvalidClientError,
+} from 'web3-common/dist/errors';
+import { WebSocketOptions, WSRequestItem } from './types';
+import { DeferredPromise } from './deferredPromise';
+
+export default class WebSocketProvider extends Web3BaseProvider {
+	private readonly wsEventEmitter: EventEmitter = new EventEmitter();
+
+	private readonly clientUrl: string;
+	private readonly wsProviderOptions: WebSocketOptions;
+
+	private webSocketConnection?: W3WS;
+
+	/* eslint-disable @typescript-eslint/no-explicit-any */
+	private readonly requestQueue: Map<JsonRpcId, WSRequestItem<any, any>>;
+	/* eslint-disable @typescript-eslint/no-explicit-any */
+	private readonly sentQueue: Map<JsonRpcId, WSRequestItem<any, any>>;
+
+	private providerStatus: Web3BaseProviderStatus;
+
+	private reconnectAttempts!: number;
+	private lastDataChunk!: string;
+	private lastChunkTimeout!: NodeJS.Timeout;
+
+	public constructor(clientUrl: string, wsProviderOptions?: WebSocketOptions) {
+		super();
+		if (!WebSocketProvider.validateProviderUrl(clientUrl))
+			throw new InvalidClientError(clientUrl);
+
+		this.clientUrl = clientUrl;
+		if (wsProviderOptions !== undefined) this.wsProviderOptions = wsProviderOptions;
+		else
+			this.wsProviderOptions = {
+				customTimeout: 1000 * 15,
+			};
+
+		if (!this.wsProviderOptions.customTimeout) this.wsProviderOptions.customTimeout = 1000 * 15;
+
+		if (!this.wsProviderOptions.reconnectOptions)
+			this.wsProviderOptions.reconnectOptions = {
+				auto: false,
+				delay: 5000,
+				maxAttempts: 5,
+				onTimeout: false,
+			};
```

Comment:
Does all this need to be defined? Because if a user provided `wsProviderOptions` as:

```typescript
{
    reconnectOptions: {}
}
```

none of this is going to get set. This also doesn't guarentee all those properties are set if the user does:

```typescript
{
    reconnectOptions: {
        delay: 5000
    }
}
```